diff -uNr munge-0.5.10.ORIG/src/etc/munge.init.in munge-0.5.10/src/etc/munge.init.in
--- munge-0.5.10.ORIG/src/etc/munge.init.in	2011-02-27 10:34:34.980855448 +0100
+++ munge-0.5.10/src/etc/munge.init.in	2011-02-27 10:36:37.645394273 +0100
@@ -115,9 +115,16 @@
 {
 # Start the service.
 #
+# If there is no key then exit as a configuration error.
+if [ ! -f /etc/munge/munge.key ] ; then
+    echo "/etc/munge/munge.key does not exist, generate with create-munge-key"
+    exit 6
+fi 
+
 # Required by LSB, where running "start" on a service already running should be
 #   considered successful.
 ##
+
   log_init "Starting $DESC" "$DAEMON_NAME"
 
   VARRUNDIR="$localstatedir/run/munge"
