diff -uNr munge-0.5.8.ORIG/src/etc/munge.init.in munge-0.5.8/src/etc/munge.init.in
--- munge-0.5.8.ORIG/src/etc/munge.init.in	2007-02-04 08:52:37.000000000 +0100
+++ munge-0.5.8/src/etc/munge.init.in	2009-06-12 11:46:40.000000000 +0200
@@ -7,7 +7,7 @@
 # Written by Chris Dunlap <cdunlap@llnl.gov>.
 # UCRL-CODE-155910.
 ###############################################################################
-# chkconfig:      345 40 60
+# chkconfig:      - 40 60
 ###############################################################################
 ### BEGIN INIT INFO
 # Provides:       munge
@@ -15,8 +15,8 @@
 # Required-Stop:  $named $time
 # Should-Start:   $local_fs $syslog
 # Should-Stop:    $local_fs $syslog
-# Default-Start:  3 4 5
 # Default-Stop:   0 1 2 6
+# Short-Description: Start/Stop the MUNGE authentication service.
 # Description:    Start/Stop the MUNGE authentication service.
 ### END INIT INFO
 ###############################################################################
